src/main/java/com/epam/aidial/core/service/PublicationService.java

```
@@ -0,0 +1,359 @@
+            }
+
+            return encodePublications(publications);
+        });
+
+        return publication;
+    }
+
+    public boolean deletePublication(ResourceDescription resource) {
+        if (resource.isFolder() || resource.getParentPath() != null) {
+            throw new IllegalArgumentException("Bad publication url: " + resource.getUrl());
+        }
+
+        MutableObject<Publication> deleted = new MutableObject<>();
+
+        resources.computeResource(PUBLIC_PUBLICATIONS, body -> {
+            Map<String, Publication> publications = decodePublications(body);
+            Publication publication = publications.remove(resource.getUrl());
+            return (publication == null) ? body : encodePublications(publications);
+        });
+
+        resources.computeResource(localPublications(resource), body -> {
+            Map<String, Publication> publications = decodePublications(body);
+            Publication publication = publications.remove(resource.getUrl());
+
+            if (publication == null) {
+                return body;
+            }
+
+            deleted.setValue(publication);
+            return encodePublications(publications);
+        });
+
+        Publication publication = deleted.getValue();
+
+        if (publication == null) {
+            return false;
+        }
+
+        if (publication.getStatus() == Publication.Status.PENDING) {
+            deleteReviewResources(publication);
+        }
+
+        return true;
+    }
+
+    private void preparePublication(ResourceDescription bucket, Publication publication) {
+        if (publication.getSourceUrl() == null) {
+            throw new IllegalArgumentException("Publication \"sourceUrl\" is missing");
+        }
+
+        if (publication.getTargetUrl() == null) {
+            throw new IllegalArgumentException("Publication \"targetUrl\" is missing");
+        }
+
+        if (publication.getResources() == null) {
+            publication.setResources(List.of());
```

probably if resources are empty we need to throw an exception

---

There is a case when you want to change just rules.

---

create operation responsible for update as well?